### ISSUE TITLE
Bug with custom fields causing N queries

### ISSUE THREAD

**Issue by rsomani95** [MrThearMan/graphene-django-query-optimizer#77]
### I have read the docs thoroughly before making this bug report.

- [X] Yes

### I have read through other open issues, and my issue is not a duplicate.

- [X] Yes

### What version of the library you are using?

0.2.8

### Which python version are you using?

3.12

### What operating system are you on?

Mac

### Description

I'm trying to define an alternate name for a given field on my model (I aim to, in my project, setup some different options / filters on this field), but am running into N queries.

Maybe I'm setting this up incorrectly?

```python
class PropertyManagerNode(IsTypeOfProxyPatch, DjangoObjectType):
    housing_companies = DjangoConnectionField(HousingCompanyNode)

    housing_companies_alt = DjangoConnectionField(HousingCompanyNode)

    @required_fields("housing_companies")
    def resolve_housing_companies_alt(model: PropertyManagerProxy, info: GQLInfo):
        return model.housing_companies.all()
```

```graphql
{
  pagedPropertyManagers {
    edges {
      node {
        name
        housingCompaniesAlt {
          edges {
            node {
              name
            }
          }
        }
      }
    }
  }
}
```

This causes N queries whereas swapping out `"housingCompaniesAlt"` for `"housingCompanies"` does the right thing (as expected).

---

I'm not sure if this is a separate bug or related to the above (I suspect related), but custom filter fields on the `HousingCompanyNode` are available in the schema, but do not work. In the example above, `HousingCompanyNode` has a custom `HousingCompanyFilterSet` class with an `address` filter defined on it:

https://github.com/MrThearMan/graphene-django-query-optimizer/blob/9d58c3be48faa6a4551f06c20c3c712dff4f7068/tests/example/types.py#L304-L347

```graphql
{
  pagedPropertyManagers {
    edges {
      node {
        name
        housingCompaniesAlt(address: "blah") {
          edges {
            node {
              name
            }
          }
        }
      }
    }
  }
}
```

Gives me this error:

```python
2024-03-08T22:42:45+0000 | INFO | project.logging.resolve:40 | Traceback (most recent call last):
  File "/Users/rahulsomani/git/graphene-django-query-optimizer/tests/project/logging.py", line 38, in resolve
    return next(root, info, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/rahulsomani/git/graphene-django-query-optimizer/query_optimizer/fields.py", line 210, in connection_resolver
    result = self.resolver(root, info, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: PropertyManagerNode.resolve_housing_companies_alt() got an unexpected keyword argument 'address'
```

**Comment by MrThearMan** [comment]
A couple of things here: 

First, `related_fields` is not meant to be used for "relation fields", i.e., ForeignKeys, ManyToManyFields, and the like. When the OptimizationCompiler finds to the hinted fields, it adds them to [optimizer.only_fields](https://github.com/MrThearMan/graphene-django-query-optimizer/blob/932e98ae711b0485c97778cb90cdba2a4458d2cf/query_optimizer/compiler.py#L431), and not the `select_related` or `prefetch_related` dictionaries, meaning the correct joins won't be performed by the Optimizer. It can be used for related fields, e.g., `@required_fields("housing_companies__name")`, in which case only that field is fetched for the model.

In fact, the whole mechanism isn't designed for allowing "alternate fields", as you put it. Even if we made the join correctly, any subsequent joins indicated by the query would not happen, since the optimizer won't continue optmization further from a hint.

Secondly, while looking at this problem, I noticed that the hinting system was unable to find hints from `graphene.Field` resolvers. I fixed for this in [0.2.11](https://github.com/MrThearMan/graphene-django-query-optimizer/releases), and also added a `required_relations` decorator that is able to handle the aforementioned "relation fields" correctly, BUT it still won't optimize past the hint you have provided, and also does not allow hinting further relations past the current model. 

Rather than try to patch the current mechanism any further than this, I would like to think about a more comprehensive overhaul for the hinting system, if I decide to support alternate fields. Suggestions are welcome.

Lastly, the error you are receiving is related to the resolver you have specified. It doesn't accept `**kwargs`, so there is not keyword to bind the filter to. This is part of the regular resolver mechanism in `graphene_django`. The filtering should probably be handled correctly by the connection field, but I'm not sure without trying first, and I'm not making guarantees at the moment that alternate fields will behave correctly.

**Comment by MrThearMan** [comment]
> ...I'm thinking if it's somehow possible to use the model fields logic itself instead of having to define a separate `.relations` which needs to be managed in parallel...

While this is likely possible, I don't think it's not the best solution for your problem, since the whole point of the optimizer is that it is able to compile and perform only the required database queries in the given request. If we hand the resposibility to the resolver, then any optimizations made there will be one-size-fits-all, and you'll likely be over or under-fetching data from the database with different queries. You'll also need to re-implement a lot of things the optimizer takes care of, like filtering and pagination.

Instead, the idea of "alternate fields" leads me into thinking that we need some mechanism of pointing to the optimizer "hey, optimize this field as if it was this relation on the model". Something like `@field_alias("housing_companies")` annotation on the resolver.

Then again, if we're just redirecing the optimizer to the relation, why would we need to define a resolver for it? Maybe a parameter to `DjangoConnectionField` is all you would need for that: `DjangoConnectionField(HousingCompanyNode, related_name="housing_companies")`. In the original design of the optimizer, I avoided adding parameters to fields, but after the `0.2.0` rewrite, the library required the use of custom fields anyway, so this woudn't be an issue.

Also, I think any solution for this will replace `@required_relations` and fetching related fields (e.g., `property_manager__name`) in `@required_fields`, since it should make those obsolete.

**Comment by MrThearMan** [comment]
I've made a first iteration of a possible solution to the [annotation-replacement](https://github.com/MrThearMan/graphene-django-query-optimizer/tree/annotation-replacement) branch. The notable API changes are as follows:

`@required_annotations` is replaced with `AnnotatedField`. It can be used like this:

```python
class HousingCompanyType(DjangoObjectType):

    greeting = AnnotatedField(graphene.String, F("name"))

    # Optional resolver. If not defined, will return `root.name`
    def resolve_greeting(root: HousingCompany, info: GQLInfo) -> str:
        return f"Hello {root.name}!"
```

This is basically the same as `@required_annotations`, but simpler.

`@required_fields` and `@required_relations` are replaced with the `field_name` argument to `RelatedField`, `DjangoListField`, and `DjangoConnectionField`. It can be used like this:

```python
class HousingCompanyNode(IsTypeOfProxyPatch, DjangoObjectType):
    developers_alt = DjangoListField(DeveloperNode, field_name="developers")
```

This will tell the optimizer that it should treat the field as if it was connected to "develpers", and use a `Prefetch(..., to_attr="developers_alt")` for prefetches. This allows querying any fields from the "alternate field" as if it was the original one, and the optimizer will also optimize its related entities. 

`RelatedFields` _seems_ to work, but I'll need to check more on them (could be clashing if the original field if that is also queried, may need to promote these to prefetches). I also want to think of more on what the annotations enable that these don't, or what else should be included with this update.

**Comment by MrThearMan** [comment]
Desided to replace the annotations with the above mechanisms. They seem more useful than the annotations in general, and if any use cases can not be replicated, I'll think of solutions then. Released on [0.4.0](https://github.com/MrThearMan/graphene-django-query-optimizer/releases/tag/v0.4.0).

**Comment by rsomani95** [comment]
@MrThearMan sorry for the delay, i got pulled in a different direction, and only just had a chance to look at this. 

This is an awesome fix for the "alternate fields". 

---

However, I have a few concerns (questions, really) around how to reformulate what `required_fields` and `annotated_fields` enabled for me before. 

---

### `required_fields`

In this simplified example, I have a computed property defined on the Django model directly, like so:

```python
class RootModel:
    bucket = models.CharField()
    key = models.CharField()

    @property
    def params(self):
        return {"Bucket": self.bucket, "Key": self.key}

    @property
    def url(self):
        cache_key = self.cache_key
        params = self.signed_url_params

        # Save to cache if not in cache, else fetch from cache
```

For which the `@required_fields` worked really well. I simply added the annotation and it did everything in one query.
```python
class UrlMixin(DjangoObjectType):
    @required_fields("key", "bucket")
    def resolve_signed_url(root: RootModel, info):
        return root.url
```

I'm not sure how to do this with `AnnotatedField` though.

---

### `annotated_fields`

An `AnnotatedField` currently only takes a single expression, whereas previously we could pass multiple `**kwargs` to `@required_annotations`.

Earlier, I had an annotated field like so:

```python
class Blah(DjangoObjectType):
    @required_annotations(
        in_time_seconds=Cast(F("in_time"), models.FloatField()) / F("in_time_base"),
        out_time_seconds=Cast(F("out_time"), models.FloatField()) / F("out_time_base"),
        duration=F("out_time_seconds") - F("in_time_seconds"),
    )
    def resolve_duration(root: TModel, info: GQLInfo):
        return root.duration
```

which I think isn't possible with this current design (I don't think the above can be formulated into a single `expression`). Maybe `AnnotatedField` can accept either one, or a list of expressions, and work accordingly. Thoughts?

**Comment by MrThearMan** [comment]
I opened another issue for re-enabling these with the new model, let's move discussion there.
